sklearn/decomposition/tests/test_pca.py

```diff
@@ -321,31 +425,85 @@ def test_pca_score2():
     n, p = 100, 3
     rng = np.random.RandomState(0)
     X = rng.randn(n, p) * .1 + np.array([3, 4, 5])
-    pca = PCA(n_components=2)
-    pca.fit(X)
-    ll1 = pca.score(X)
-    ll2 = pca.score(rng.randn(n, p) * .2 + np.array([3, 4, 5]))
-    assert_greater(ll1, ll2)
+    for solver in solver_list:
+        pca = PCA(n_components=2, svd_solver=solver)
+        pca.fit(X)
+        ll1 = pca.score(X)
+        ll2 = pca.score(rng.randn(n, p) * .2 + np.array([3, 4, 5]))
+        assert_greater(ll1, ll2)
 
-    # Test that it gives the same scores if whiten=True
-    pca = PCA(n_components=2, whiten=True)
-    pca.fit(X)
-    ll2 = pca.score(X)
-    assert_almost_equal(ll1, ll2)
+        # Test that it gives different scores if whiten=True
+        pca = PCA(n_components=2, whiten=True, svd_solver=solver)
+        pca.fit(X)
+        ll2 = pca.score(X)
+        assert_true(ll1 != ll2)
 
 
 def test_pca_score3():
     # Check that probabilistic PCA selects the right model
     n, p = 200, 3
     rng = np.random.RandomState(0)
-    Xl = (rng.randn(n, p) + rng.randn(n, 1) * np.array([3, 4, 5])
-          + np.array([1, 0, 7]))
-    Xt = (rng.randn(n, p) + rng.randn(n, 1) * np.array([3, 4, 5])
-          + np.array([1, 0, 7]))
+    Xl = (rng.randn(n, p) + rng.randn(n, 1) * np.array([3, 4, 5]) +
+          np.array([1, 0, 7]))
+    Xt = (rng.randn(n, p) + rng.randn(n, 1) * np.array([3, 4, 5]) +
+          np.array([1, 0, 7]))
     ll = np.zeros(p)
     for k in range(p):
-        pca = PCA(n_components=k)
+        pca = PCA(n_components=k, svd_solver='full')
         pca.fit(Xl)
         ll[k] = pca.score(Xt)
 
     assert_true(ll.argmax() == 1)
+
+
+def test_svd_solver_auto():
+    rng = np.random.RandomState(0)
+    X = rng.uniform(size=(1000, 50))
+
+    # case: n_components in (0,1) => 'full'
+    pca = PCA(n_components=.5)
+    pca.fit(X)
+    pca_test = PCA(n_components=.5, svd_solver='full')
+    pca_test.fit(X)
+    assert_array_almost_equal(pca.components_, pca_test.components_)
+
+    # case: max(X.shape) <= 500 => 'full'
+    pca = PCA(n_components=5, random_state=0)
+    Y = X[:10, :]
+    pca.fit(Y)
+    pca_test = PCA(n_components=5, svd_solver='full', random_state=0)
+    pca_test.fit(Y)
+    assert_array_almost_equal(pca.components_, pca_test.components_)
+
+    # case: n_components >= .8 * min(X.shape) => 'full'
+    pca = PCA(n_components=50)
+    pca.fit(X)
+    pca_test = PCA(n_components=50, svd_solver='full')
+    pca_test.fit(X)
+    assert_array_almost_equal(pca.components_, pca_test.components_)
+
+    # n_components >= 1 and n_components < .8 * min(X.shape) => 'randomized'
+    pca = PCA(n_components=10, random_state=0)
+    pca.fit(X)
+    pca_test = PCA(n_components=10, svd_solver='randomized', random_state=0)
+    pca_test.fit(X)
+    assert_array_almost_equal(pca.components_, pca_test.components_)
```

Comment:
nitpick: Shouldn't these be exactly the same, considering you are doing the same thing?